cpplint.py

```diff
@@ -934,9 +934,93 @@ def unicode_escape_decode(x):
 # This is set by --headers flag.
 _hpp_headers = set([])
 
-# {str, bool}: a map from error categories to booleans which indicate if the
-# category should be suppressed for every line.
-_global_error_suppressions = {}
+class ErrorSuppressions:
+  """Class to track all error suppressions for cpplint"""
+
+  class SuppressionRange:
+    """Class to represent a range of line numbers for which an error is suppressed"""
+    def __init__(self, begin=0, end=math.inf):
+      self._begin = begin
+      self._end = end
+
+    def __str__(self):
+      return f'[{self.begin}-{self.end}]'
+
+    def __contains__(self, obj):
+      if isinstance(obj, ErrorSuppressions.SuppressionRange):
+        return self.begin <= obj.begin and self.end >= obj.end
+      elif isinstance(obj, int):
+        return self.begin <= obj <= self.end
+      raise TypeError()
+
+    @property
+    def begin(self):
+      return self._begin
+
+    @property
+    def end(self):
+      return self._end
+
+    @end.setter
+    def end(self, end):
+      assert end > self.begin
+      self._end = end
+
+  def __init__(self):
+    self._suppressions = {}
+    self._open_block_suppressions = []
+
+  def _add_suppression(self, category, suppression_range):
+    suppressed = self._suppressions.setdefault(category, [])
+    if not suppressed or suppression_range not in suppressed[-1]:
+      suppressed.append(suppression_range)
+      return True
+    return False
+
+  @property
+  def open_block_start(self):
+    """:return: The start of the current open block or `-1` if there is not an open block"""
+    return self._open_block_suppressions[0].begin if self._open_block_suppressions else -1
+
+  def add_global_suppression(self, category):
+    """Add a suppression for `category` which is suppressed for the whole file"""
+    self._suppressions[category] = (self.SuppressionRange(),)
+
+  def add_line_suppression(self, category, linenum):
+    """Add a suppression for `category` which is suppressed only on `linenum`"""
+    self._add_suppression(category, self.SuppressionRange(linenum, linenum))
+
+  def start_block_suppression(self, category, linenum):
+    """Start a suppression block for `category` on `linenum`. inclusive"""
+    suppression_range = self.SuppressionRange(linenum)
+    if self._add_suppression(category, suppression_range):
+      self._open_block_suppressions.append(suppression_range)
+
+  def end_block_suppression(self, linenum):
+    """End the current block suppression on `linenum`. inclusive"""
+    for suppression in self._open_block_suppressions:
+      suppression.end = linenum
+    self._open_block_suppressions.clear()
+
+  def is_suppressed(self, category, linenum):
+    """:return: `True` if `category` is suppressed for `linenum`"""
+    suppressed = self._suppressions.get(category) or self._suppressions.get(None)
```

Comment:
Not sure if this is a bug, but if we check category "foo" for linenum 42, and we have a line suppression for "foo" for line 13, and an open block for None going from lines 40-inf, will we miss the second and treat this as unsupressed?

Reply:
nit: I Would also accept `self._suppressions.get(None, [])` to avoid one `if` below

Reply:
I fixed this so that the lists are appended and checked